src/sync/sharded_lock.rs

```diff
@@ -0,0 +1,149 @@
+//! traditional reader-writer locks.
+
+use std::cell::UnsafeCell;
+use std::marker::PhantomData;
+use std::mem;
+use std::ops::{Deref, DerefMut};
+
+use num_cpus;
+use parking_lot::{RwLock, RwLockReadGuard, RwLockWriteGuard};
+
+use CachePadded;
+
+/// A scalable reader-writer lock.
+///
+/// This type of lock allows a number of readers or at most one writer at any point in time. The
+/// write portion of this lock typically allows modification of the underlying data (exclusive
+/// access) and the read portion of this lock typically allows for read-only access (shared
+/// access).
+///
+/// This reader-writer lock differs from typical implementations in that it internally creates a
+/// list of reader-writer locks called 'shards'. Shards are aligned and padded to the cache line
+/// size.
+///
+/// Read operations lock only one shard specific to the current thread, while write operations lock
+/// every shard in succession. This strategy makes concurrent read operations faster due to less
+/// contention, but write operations are slower due to increased amount of locking.
+pub struct ShardedLock<T> {
+    /// A list of locks protecting the internal data.
+    shards: Vec<CachePadded<RwLock<()>>>,
+
+    /// The internal data.
+    value: UnsafeCell<T>,
+}
+
+unsafe impl<T: Send> Send for ShardedLock<T> {}
+unsafe impl<T: Send + Sync> Sync for ShardedLock<T> {}
+
+impl<T> ShardedLock<T> {
+    /// Creates a new `ShardedLock` initialized with `value`.
+    pub fn new(value: T) -> ShardedLock<T> {
+        // The number of shards is a power of two so that the modulo operation in `read` becomes a
+        // simple bitwise "and".
+        let num_shards = num_cpus::get().next_power_of_two();
+
+        ShardedLock {
+            shards: (0..num_shards)
+                .map(|_| CachePadded::new(RwLock::new(())))
+                .collect(),
+            value: UnsafeCell::new(value),
+        }
+    }
+
+    /// Locks with shared read access, blocking the current thread until it can be acquired.
+    ///
+    /// The calling thread will be blocked until there are no more writers which hold the lock.
+    /// There may be other readers currently inside the lock when this method returns. This method
+    /// does not provide any guarantees with respect to the ordering of whether contentious readers
+    /// or writers will acquire the lock first.
+    ///
+    /// Returns an RAII guard which will drop the read access of this lock when dropped.
+    pub fn read(&self) -> ShardedLockReadGuard<T> {
+        // Take the current thread index and map it to a shard index. Thread indices will tend to
+        // distribute shards among threads equally, thus reducing contention due to read-locking.
+        let current_index = {
+            #[cfg(feature = "use_std")] { ::thread::current_index().unwrap_or(0) }
+            #[cfg(not(feature = "use_std"))] 0
+        };
```

Comment:
`parking_lot` requires std, so the `cfg` here is pointless. You should instead `#[cfg]` out the entire module for no_std builds.